pkg/v1/remote/write.go

```diff
@@ -120,52 +120,79 @@ func (w *writer) nextLocation(resp *http.Response) (string, error) {
 	return resp.Request.URL.ResolveReference(u).String(), nil
 }
 
+// checkExisting checks if a blob exists already in the repository by making a
+// HEAD request to the blob store API.
+func (w *writer) checkExisting(h v1.Hash) (bool, error) {
+	u := w.url(fmt.Sprintf("/v2/%s/blobs/%s", w.ref.Context().RepositoryStr(), h.String()))
+
+	resp, err := w.client.Head(u.String())
+	if err != nil {
+		return false, err
+	}
+	defer resp.Body.Close()
+
+	if err := CheckError(resp, http.StatusOK, http.StatusNotFound); err != nil {
+		return false, err
+	}
+
+	if resp.StatusCode == http.StatusOK {
+		return true, nil
+	}
+
+	return false, nil
+}
+
 // initiateUpload initiates the blob upload, which starts with a POST that can
 // optionally include the hash of the layer and a list of repositories from
 // which that layer might be read. On failure, an error is returned.
 // On success, the layer was either mounted (nothing more to do) or a blob
 // upload was initiated and the body of that blob should be sent to the returned
 // location.
 func (w *writer) initiateUpload(h v1.Hash) (location string, mounted bool, err error) {
-	u := w.url(fmt.Sprintf("/v2/%s/blobs/uploads/", w.ref.Context().RepositoryStr()))
+	l, err := w.img.LayerByDigest(h)
+	if err != nil {
+		return
+	}
+
 	uv := url.Values{
 		"mount": []string{h.String()},
 	}
-	l, err := w.img.LayerByDigest(h)
-	if err != nil {
-		return "", false, err
-	}
-	// We currently avoid HEAD because it's semi-redundant with the mount that is part
-	// of initiating the blob upload.  GCR will perform an existence check on the initiation
-	// if "mount" is specified, even if no "from" sources are specified.  If this turns out
-	// to not be broadly applicable then we should replace mounts without "from"s with a HEAD.
-	if ml, ok := l.(*MountableLayer); ok {
-		if w.ref.Context().RegistryStr() == ml.Reference.Context().RegistryStr() {
-			uv["from"] = []string{ml.Reference.Context().RepositoryStr()}
+	// GCR will perform an existence check on the initiation if "mount" is
+	// specified, even if no "from" sources are specified.  However, this is not
+	// broadly applicable to all registries, e.g. ECR. So we use a simple HEAD
+	// check when the "from" source is not specified.
+	if ml, ok := l.(*MountableLayer); ok && w.ref.Context().RegistryStr() == ml.Reference.Context().RegistryStr() {
+		uv["from"] = []string{ml.Reference.Context().RepositoryStr()}
+	} else {
```

Comment:
It seems like this misses the case where you're doing an ECR -> ECR upload, though maybe I'm misreading it. (e.g. `crane copy <ecr-src> <ecr-dst>` to retag something would not end up calling `checkExisting`)

Reply:
Yes, you are right. I just went off the comment there and implemented what it said. But just checked, it did not work for `crane copy`. Now fixed; crane copy to an ECR destination will now reuse blobs in the destination image repository, if they exist.